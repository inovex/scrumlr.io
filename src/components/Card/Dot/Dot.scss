@import '../../../constants/style';

$dot-color: $theme-color-primary;

.dot__small {
  width: 28px;
  height: 28px;
  line-height: 30px;
}

.dot__large {
  width: 32px;
  height: 32px;
  line-height: 32px;
}

.dot {
  display: block;
  text-align: center;
  border: 0;
  margin: 0;
  padding: 0;
  background: $dot-color;
  color: $theme-color-secondary;
  border-radius: 100%;
  position: relative;
  outline: none;
}

.dot__interactive {
  cursor: pointer;
}

.dot__folded-corner {
  width: 0;
  height: 0;
  position: absolute;
  top: 0;
  right: 0;
  border-width: 0;
  border-style: solid;
  background: hsla(260, 5%, 75%, 0.5);
  box-shadow: 2px -2px 0px #fff, 0 1px 1px rgba(0, 0, 0, 0.1),
    -1px 1px 1px rgba(0, 0, 0, 0.1);
  border-radius: 0 0 0 16px;
  border-color: $theme-color-background $theme-color-background transparent
    transparent;
  @include animate(border-width);
}

.dot.dot__small:hover .dot__folded-corner {
  border-width: 12px 0 0 12px;
}

.dot.dot__large:hover .dot__folded-corner {
  border-width: 18px 0 0 18px;
}
