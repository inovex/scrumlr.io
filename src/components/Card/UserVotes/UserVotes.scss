@import '../../../constants/style';

$icon__width: 24px;
$dot-color: #3991f4;

.user-votes {
  display: flex;
  align-items: center;
}

.user-votes__dots {
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.user-votes__dots-item {
  display: inline-block;
}
