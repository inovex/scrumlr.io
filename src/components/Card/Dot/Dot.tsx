import * as classNames from 'classnames';
import * as React from 'react';

import './Dot.scss';

export type DotSize = 'small' | 'large';

export interface DotProps {
  size: DotSize;
  animate?: boolean;
  onClick?: (() => void) | undefined;

  [key: string]: any;
}

export class Dot extends React.Component<DotProps, {}> {
  static defaultProps: Partial<DotProps> = {
    animate: false,
    onClick: undefined
  };

  renderContent = () => {
    const { children, animate, onClick, size } = this.props;

    if (onClick) {
      return (
        <button
          type="button"
          className={classNames('dot', 'dot__interactive', `dot__${size}`)}
          onClick={onClick}
        >
          {animate && <span className="dot__folded-corner" />}

          {children}
        </button>
      );
    }

    return <div className={classNames('dot', `dot__${size}`)}>{children}</div>;
  };

  render() {
    const { className, size } = this.props;

    return (
      <span className={classNames(`dot__${size}`, className)}>
        {this.renderContent()}
      </span>
    );
  }
}

export default Dot;
