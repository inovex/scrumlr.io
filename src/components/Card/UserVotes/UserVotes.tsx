import * as React from 'react';
import { range } from 'lodash';

import { AddDot, DownvoteDot } from '../Dot';
import './UserVotes.scss';

export interface UserVotesProps {
  onDownvote: () => void;
  onUpvote: () => void;
  votes?: number;
  maxDotsShown?: number;
}

const defaultProps: Partial<UserVotesProps> = {
  votes: 0,
  maxDotsShown: 5
};

export class UserVotes extends React.Component<UserVotesProps, {}> {
  static defaultProps: Partial<UserVotesProps> = defaultProps;

  renderDownvote(votes?: number) {
    return (
      <DownvoteDot onClick={() => this.props.onDownvote()}>{votes}</DownvoteDot>
    );
  }

  renderDots() {
    const { votes = 0, maxDotsShown } = this.props;

    if (maxDotsShown && votes && votes > maxDotsShown) {
      return this.renderDownvote(votes);
    }

    return (
      <ul className="user-votes__dots">
        {range(votes).map((x, i) => (
          <li key={votes - i} className="user-votes__dots-item">
            {this.renderDownvote()}
          </li>
        ))}
      </ul>
    );
  }

  render() {
    return (
      <span className="user-votes">
        {this.renderDots()}
        <AddDot onClick={() => this.props.onUpvote()} />
      </span>
    );
  }
}

export default UserVotes;
