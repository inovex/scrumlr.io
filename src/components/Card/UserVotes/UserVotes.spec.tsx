import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { AddDot, DownvoteDot } from '../Dot';

import { UserVotes, UserVotesProps } from './UserVotes';

describe('<UserVotes />', () => {
  let wrapper: ShallowWrapper<UserVotesProps, {}>;
  let props: UserVotesProps = {
    votes: 0,
    onDownvote: jest.fn(),
    onUpvote: jest.fn()
  };

  it('should have default values', () => {
    expect(UserVotes.defaultProps).toMatchSnapshot();
  });

  it('should as many dots as votes are available', () => {
    wrapper = shallow(<UserVotes {...props} votes={0} />);
    let dots = wrapper.find(DownvoteDot);
    expect(dots).toHaveLength(0);

    wrapper = shallow(<UserVotes {...props} votes={3} />);
    dots = wrapper.find(DownvoteDot);
    expect(dots).toHaveLength(1);
  });

  it('should render only one dot if more votes are available than maximal allowed dots', () => {
    const votes = 5;
    wrapper = shallow(<UserVotes {...props} votes={votes} />);
    const dots = wrapper.find(DownvoteDot);
    expect(dots).toHaveLength(1);
    expect(dots.prop('children')).toEqual(votes);
  });

  it('should call onDownvote callback if a DownvoteDot is clicked', () => {
    wrapper = shallow(<UserVotes {...props} votes={1} />);
    const dots = wrapper.find(DownvoteDot);
    expect(props.onDownvote).not.toHaveBeenCalled();
    dots.at(0).simulate('click');
    expect(props.onDownvote).toHaveBeenCalled();
  });

  it('should render an AddDot component', () => {
    wrapper = shallow(<UserVotes {...props} />);
    const addDot = wrapper.find(AddDot);
    expect(addDot).toHaveLength(1);
  });

  it('should call onUpvote callback if the AddDot component is clicked', () => {
    wrapper = shallow(<UserVotes {...props} />);
    const addDot = wrapper.find(AddDot);
    expect(props.onUpvote).not.toHaveBeenCalled();
    addDot.at(0).simulate('click');
    expect(props.onUpvote).toHaveBeenCalled();
  });
});
