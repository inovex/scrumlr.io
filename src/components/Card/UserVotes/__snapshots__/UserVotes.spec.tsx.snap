// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserVotes /> should have default values 1`] = `
Object {
  "votes": 0,
}
`;
